pkg/skaffold/build/result.go

```diff
@@ -0,0 +1,152 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package build
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+)
+
+const bufferedLinesPerArtifact = 10000
+
+// For testing
+var (
+	buffSize = bufferedLinesPerArtifact
+)
+
+// logAggregator provides an interface to create an output writer for each artifact build and later aggregate the logs in build order.
+// The order of output is not guaranteed between multiple builds running concurrently.
+type logAggregator interface {
+	// GetWriter returns an output writer tracked by the logAggregator
+	GetWriter() (io.WriteCloser, error)
+	// PrintInOrder prints the output from each allotted writer in build order.
+	// It blocks until the instantiated capacity of io writers have been all allotted and closed, or the context is cancelled.
+	PrintInOrder(ctx context.Context, out io.Writer)
+}
+
+type logAggregatorImpl struct {
+	messages   chan chan string
+	size       int
+	capacity   int
+	countMutex sync.Mutex
+}
+
+func (l *logAggregatorImpl) GetWriter() (io.WriteCloser, error) {
+	if err := l.checkCapacity(); err != nil {
```

Comment:
this condition would never happen as channel capacity is size of artifacts. Can we drop this condition to make things simpler.

Reply:
I'd like to keep it. I think that `logAggregator` can be reused somewhere else if we have a need for concurrent runs for something. So I wanted to keep it safe, otherwise there won't be any error messages if writer requests exceed the buffer capacity and the code would deadlock.